examples/triangle_intersection/triangle_intersection.cpp

```diff
@@ -287,61 +142,60 @@ class TriangleIntersectionCallback
-                            beta_coeff};
-#endif
-    bool intersects = coeffs[0] >= 0 && coeffs[1] >= 0 && coeffs[2] >= 0;
-
-    if (intersects)
-    {
-      _offsets(query_index) = triangle_index;
-      _coefficients(query_index) = coeffs;
-      return ArborX::CallbackTreeTraversalControl::early_exit;
-    }
-    return ArborX::CallbackTreeTraversalControl::normal_continuation;
+    Kokkos::Array<float, 3> coeffs = {1 - alpha_coeff - beta_coeff, alpha_coeff,
+                                      beta_coeff};
+
+    _offsets(query_index) = triangle_index;
+    _coefficients(query_index) = coeffs;
+
+    return ArborX::CallbackTreeTraversalControl::early_exit;
   }
 
-private:
-  Triangles<MemorySpace> _triangles;
   Kokkos::View<int *, MemorySpace> _offsets;
-  Kokkos::View<ArborX::Point *, MemorySpace> _coefficients;
+  Kokkos::View<Kokkos::Array<float, 3> *, MemorySpace> _coefficients;
 };
 
 // Now that we have encapsulated the objects and queries to be used within the
 // Triangles class, we can continue with performing the actual search.
-int main()
+int main(int argc, char *argv[])
 {
-  Kokkos::ScopeGuard guard;
+  Kokkos::ScopeGuard guard(argc, argv);
 
   using ExecutionSpace = Kokkos::DefaultExecutionSpace;
   using MemorySpace = typename ExecutionSpace::memory_space;
-  ExecutionSpace execution_space;
+  ExecutionSpace space;
+
+  constexpr float Lx = 50.f;
+  constexpr float Ly = 50.f;
+  constexpr int nx = 101;
+  constexpr int ny = 101;
 
   // Create grid with triangles
-  Triangles<MemorySpace> triangles(execution_space);
+  Kokkos::View<Point *, MemorySpace> points("Example::points", 0);
+  buildPoints(space, Lx, Ly, nx, ny, points);
+
+  int const num_points = points.size();
+
+  Kokkos::View<Triangle *, MemorySpace> triangles("Example::triangles", 0);
+  buildTriangles(space, Lx, Ly, nx, ny, triangles);
 
   // Create BVH tree
   ArborX::BoundingVolumeHierarchy<MemorySpace,
-                                  ArborX::PairValueIndex<Box>> const
-      tree(execution_space, ArborX::Experimental::attach_indices(triangles));
-
-  // Create the points used for queries
-  Points<MemorySpace> points(execution_space);
+                                  ArborX::PairValueIndex<Triangle>> const
+      tree(space, ArborX::Experimental::attach_indices(triangles));
 
   // Execute the queries
-  int const n = points.size();
-  Kokkos::View<int *, MemorySpace> offsets("Example::offsets", n);
-  Kokkos::View<ArborX::Point *, MemorySpace> coefficients(
-      "Example::coefficients", n);
+  Kokkos::View<int *, MemorySpace> offsets("Example::offsets", num_points);
+  Kokkos::View<Kokkos::Array<float, 3> *, MemorySpace> coefficients(
+      "Example::coefficients", num_points);
 
-  tree.query(execution_space, points,
-             TriangleIntersectionCallback<MemorySpace>{triangles, offsets,
-                                                       coefficients});
+  tree.query(space, Points<MemorySpace>{points},
```

Comment:
If we are using attach_indices for the primitives/values, then  let's get rid of Points and use it for the predicates as well.

Reply:
I'm not sure what you mean. You would still need to construct the predicates on the fly somewhere, so you still need some struct to run access traits on.